include/deal.II/grid/data_transfer.h

```diff
@@ -0,0 +1,235 @@
+   * consists of the stem @p filename and an attached identifier
+   * <tt>_fixed.data</tt> for fixed size data and <tt>_variable.data</tt>
+   * for variable size data.
+   *
+   * If MPI support is enabled, all processors write into these files
+   * simultaneously via MPIIO. Each processor's position to write to will be
+   * determined from the provided input parameters.
+   *
+   * Data has to be previously packed with pack_data().
+   */
+  void
+  save(const unsigned int global_first_cell,
+       const unsigned int global_num_cells,
+       const std::string &filename) const;
+
+  /**
+   * Transfer data from file system.
+   *
+   * The data will be read from separate file, whose name
+   * consists of the stem @p filename and an attached identifier
+   * <tt>_fixed.data</tt> for fixed size data and <tt>_variable.data</tt>
+   * for variable size data.
+   * The @p n_attached_deserialize_fixed and @p n_attached_deserialize_variable
+   * parameters are required to gather the memory offsets for each
+   * callback.
+   *
+   * If MPI support is enabled, all processors read from these files
+   * simultaneously via MPIIO. Each processor's position to read from will be
+   * determined from the provided input arguments.
+   *
+   * After loading, unpack_data() needs to be called to finally
+   * distribute data across the associated triangulation.
+   */
+  void
+  load(const unsigned int global_first_cell,
+       const unsigned int global_num_cells,
+       const unsigned int local_num_cells,
+       const std::string &filename,
+       const unsigned int n_attached_deserialize_fixed,
+       const unsigned int n_attached_deserialize_variable);
+
+  /**
+   * Clears all containers and associated data, and resets member
+   * values to their default state.
+   *
+   * Frees memory completely.
+   */
+  void
+  clear();
+
+  /**
+   * Flag that denotes if variable size data has been packed.
+   */
+  bool variable_size_data_stored;
+
+  /**
+   * Cumulative size in bytes that those functions that have called
+   * register_data_attach() want to attach to each cell. This number
+   * only pertains to fixed-sized buffers where the data attached to
+   * each cell has exactly the same size.
+   *
+   * The last entry of this container corresponds to the data size
+   * packed per cell in the fixed size buffer (which can be accessed
+   * calling <tt>sizes_fixed_cumulative.back()</tt>).
+   */
+  std::vector<unsigned int> sizes_fixed_cumulative;
+
+  /**
+   * Consecutive buffers designed for the fixed size transfer
+   * functions.
+   */
+  std::vector<char> src_data_fixed;
+  std::vector<char> dest_data_fixed;
+
+  /**
+   * Consecutive buffers designed for the variable size transfer
+   * functions.
+   */
+  std::vector<int>  src_sizes_variable;
+  std::vector<int>  dest_sizes_variable;
+  std::vector<char> src_data_variable;
+  std::vector<char> dest_data_variable;
+
+private:
+  const MPI_Comm &mpi_communicator;
```

Comment:
While you're here, let's make this a non-reference. We've made similar changes elsewhere in recent times.